### ISSUE TITLE
Time cost for each training step increases with the training procedure

### ISSUE THREAD

**Issue by ruiann** [tensorflow/tensorflow#9471]
When I try to run a SRGAN network by 32 images with 96 * 96 size, each training step the time cost increases. At the beginning, each step cost 35 seconds, but when 160 steps later, the time cost increases to more than 200 seconds. By checking the time log, I can see it do increase with the training step. If I save the model and restart training process, the time cost reduces to about 35 seconds and start increasing again.

**Comment by aselle** [comment]
Please provide details about what platform you are using  (operating system, architecture). Also include your TensorFlow version. Also, did you compile from source or install a binary?  Make sure you also include the exact command if possible to produce  the output included in your test case. If you are unclear what to include  see the issue template displayed in  [the Github new issue template](https://github.com/tensorflow/tensorflow/issues/new).

 We ask for this in the issue submission template, because    it is really difficult to help without that information. Thanks!

**Comment by ruiann** [comment]
== cat /etc/issue ===============================================
Linux inverse 4.4.0-75-generic #96-Ubuntu SMP Thu Apr 20 09:56:33 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
VERSION="16.04.2 LTS (Xenial Xerus)"
VERSION_ID="16.04"
VERSION_CODENAME=xenial

== are we in docker =============================================
No

== compiler =====================================================
c++ (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


== uname -a =====================================================
Linux inverse 4.4.0-75-generic #96-Ubuntu SMP Thu Apr 20 09:56:33 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

== check pips ===================================================
numpy (1.12.1rc1)
protobuf (3.2.0)
tensorflow-gpu (1.0.1)

== check for virtualenv =========================================
False

== tensorflow import ============================================
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcublas.so.8.0 locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcudnn.so.5 locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcufft.so.8.0 locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcurand.so.8.0 locally
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE3 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:910] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
I tensorflow/core/common_runtime/gpu/gpu_device.cc:885] Found device 0 with properties: 
name: GeForce GTX 1060 6GB
major: 6 minor: 1 memoryClockRate (GHz) 1.7845
pciBusID 0000:01:00.0
Total memory: 5.92GiB
Free memory: 1.19GiB
I tensorflow/core/common_runtime/gpu/gpu_device.cc:906] DMA: 0 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:916] 0:   Y 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:975] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 1060 6GB, pci bus id: 0000:01:00.0)
tf.VERSION = 1.0.1
tf.GIT_VERSION = v1.0.0-65-g4763edf-dirty
tf.COMPILER_VERSION = v1.0.0-65-g4763edf-dirty
Sanity check: array([1], dtype=int32)
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcublas.so.8.0 locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcudnn.so.5 locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcufft.so.8.0 locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:135] successfully opened CUDA library libcurand.so.8.0 locally

== env ==========================================================
LD_LIBRARY_PATH /usr/local/cuda-8.0/lib64
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
Thu Apr 27 21:44:49 2017       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 375.39                 Driver Version: 375.39                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 106...  Off  | 0000:01:00.0      On |                  N/A |
| 15%   46C    P2    29W / 150W |   4781MiB /  6064MiB |      3%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0      1432    G   /usr/lib/xorg/Xorg                             209MiB |
|    0      2030    G   compiz                                          62MiB |
|    0      3119    C   python                                        4427MiB |
|    0      4590    G   /home/inverse/.steam/ubuntu12_32/steam          28MiB |
|    0      6412    G   ...el-token=66A5E1EB1D60EC20A68FDE459A10598F    48MiB |
+-----------------------------------------------------------------------------+

== cuda libs  ===================================================
/usr/local/cuda-8.0/lib64/libcudart_static.a
/usr/local/cuda-8.0/lib64/libcudart.so.8.0.61
/usr/local/cuda-8.0/doc/man/man7/libcudart.so.7
/usr/local/cuda-8.0/doc/man/man7/libcudart.7

step: 1160
step cost: 44s
step: 1161
step cost: 32s
step: 1162
step cost: 34s
step: 1163
step cost: 33s
step: 1164
step cost: 33s
step: 1165
step cost: 34s
step: 1166
step cost: 34s
step: 1167
step cost: 35s
step: 1168
step cost: 36s
step: 1169
step cost: 38s
step: 1170
step cost: 39s
step: 1171
step cost: 40s
step: 1172
step cost: 42s
step: 1173
step cost: 43s
step: 1174
step cost: 44s
step: 1175
step cost: 51s
step: 1176
step cost: 50s
step: 1177
step cost: 51s
step: 1178


You can see the time cost increases with step, after restart the training task it reduces to 35s again

My training code like this
```
 while queue.size().eval() > batch_size:
        start_time = time.time()
        step = global_step.eval()
        print('step: %d' % step)
        x_batch = [get_image(image.eval()) for i in xrange(batch_size)]
        summary_str, d_loss, g_loss = sess.run([summary, d_train_op, g_train_op], feed_dict={x: x_batch})
        summary_writer.add_summary(summary_str, step)

        if global_step.eval() % 20 == 0:
                checkpoint_file = os.path.join(model_dir, 'model.latest')
                saver.save(sess, checkpoint_file)

        if global_step.eval() % 100 == 0:
                summary_writer.add_run_metadata(run_metadata, 'step%03d' % step)

        sess.run(update_global_step)
        print("step cost: %ds" % (time.time() - start_time))
summary_writer.close()
```

**Comment by aselle** [comment]
Could you monitor memory usage through the steps? 
Are you using jemalloc (which is installed by default)?
If you remove all the summary code does it still slow down? 
Can you produce a minimal reproducible test case?

**Comment by ruiann** [comment]
How can I monitor the memory usage? My OS uses the default malloc algorithm of Ubuntu 16.04 LTS, which I think is jemalloc. I will try remove the summary code later. My code only contains the definition of network and loss function, and read dir to get image samples, and use them to train

**Comment by ruiann** [comment]
@aselle 
I found the time increase in the read jpeg images, when I start train session the time cost for reading 1 image from HDD is less than 1 second, but it increases with steps,after several steps it reaches more than 3 seconds and have fluctuation which can be more than 10 seconds for some images.
My read image function is as below
```
def get_image(image_path):
    file = tf.read_file(image_path)
    image = tf.image.decode_image(file, channels=channels)
    sess = tf.get_default_session()
    file, image = sess.run([file, image])
    return image
```

I log the time in the start of this function and get time cost before return. For each training step I dequeue the batch_size of image's path from random shuffle queue and read them to build a image batch, this function will be called batch_size times

The time cost for Adam Optimizer is stable

**Comment by aselle** [comment]
https://docs.python.org/2/library/resource.html
getursage

**Comment by ruiann** [comment]
@aselle 
Thx for your guide, the memory usage do increases. The log in my MBP with CPU-version tensorflow is as below

```
step: 323
read image cost: 40s
step cost: 65s
Memory usage: 5172464
step: 324
read image cost: 40s
step cost: 67s
Memory usage: 5190316
step: 325
read image cost: 40s
step cost: 65s
Memory usage: 5224076
step: 326
read image cost: 42s
step cost: 67s
Memory usage: 5243708
step: 327
read image cost: 47s
step cost: 76s
Memory usage: 5285364
```

After several steps (from check point) the memory usage start increases, I don't know why, you can see my [code](https://github.com/ruiann/SRGAN/blob/master/SRGAN/train.py)

**Comment by ali01** [comment]
Hi @ruiann, are you able to construct a minimal reproducible example of this problem? As @aselle asked, if you remove all of the summary code, does the problem still occur?

**Comment by ruiann** [comment]
@ali01 
This is the minimal reproduce I think, no useless code, use a SRGAN network. I don't know how can I minimize it without break the network.

It occurs regardless of the summary. I wonder if it's because the random shuffle queue but have no time to deal with it recent days.

**Comment by ali01** [comment]
Hey @zheng-xq, could you take a look? Thanks!

**Comment by gunan** [comment]
Ping!
Are there any updates here?

**Comment by ruiann** [comment]
@gunan sorry I'm busy with other works. This behavior can be reproduced only in my own pc, maybe some hardware issue?

**Comment by yzhwang** [comment]
Trying to reproduce this on my local machine with CUDA 8, CUDNN v5.1.10, and tensorflow v1.1.0.
Currently has run 28 steps:
Memory usage: 4043
step: 0
read image cost: 37s
step cost: 73s
Memory usage: 4811
step: 1
read image cost: 38s
step cost: 39s
Memory usage: 4811
step: 2
read image cost: 38s
step cost: 38s
Memory usage: 4811
step: 3
read image cost: 40s
step cost: 40s
Memory usage: 4811
step: 4
read image cost: 40s
step cost: 41s
Memory usage: 4811
step: 5
read image cost: 41s
step cost: 41s
Memory usage: 4811
step: 6
read image cost: 39s
step cost: 40s
Memory usage: 4811
step: 7
read image cost: 44s
step cost: 44s
Memory usage: 4811
step: 8
read image cost: 45s
step cost: 46s
Memory usage: 4811
step: 9
read image cost: 46s
step cost: 46s
Memory usage: 4811
step: 10
read image cost: 47s
step cost: 47s
Memory usage: 4811
step: 11
read image cost: 49s
step cost: 50s
Memory usage: 4811
step: 12
read image cost: 49s
step cost: 49s
Memory usage: 4811
step: 13
read image cost: 51s
step cost: 52s
Memory usage: 4811
step: 14
read image cost: 54s
step cost: 55s
Memory usage: 4811
step: 15
read image cost: 54s
step cost: 54s
Memory usage: 4811
step: 16
read image cost: 55s
step cost: 56s
Memory usage: 4811
step: 17
read image cost: 58s
step cost: 58s
Memory usage: 4811
step: 18
read image cost: 58s
step cost: 58s
Memory usage: 4811
step: 19
read image cost: 59s
step cost: 60s
Memory usage: 4811
step: 20
read image cost: 62s
step cost: 107s
Memory usage: 5884
step: 21
read image cost: 63s
step cost: 64s
Memory usage: 5884
step: 22
read image cost: 65s
step cost: 65s
Memory usage: 5884
step: 23
read image cost: 65s
step cost: 66s
Memory usage: 5884
step: 24
read image cost: 66s
step cost: 67s
Memory usage: 5884
step: 25
read image cost: 70s
step cost: 71s
Memory usage: 5884
step: 26
read image cost: 73s
step cost: 73s
Memory usage: 5884
step: 27
read image cost: 72s
step cost: 73s
Memory usage: 5884

So the per-step training time does seem to be increasing. I will let the program run a bit longer to verify this.

**Comment by yzhwang** [comment]
So at step 40, the memory usage and run time per training step is like this:
step: 40
read image cost: 96s
step cost: 151s
Memory usage: 7049

By restarting the training both memory usage and run time drop:
step: 40
read image cost: 37s
step cost: 72s
Memory usage: 4832

I will investigate on this.

**Comment by heloowird** [comment]
I found the same problem, and I fixed it by puting `tf.image.decode_image` into building graph step instead of running step.

**Comment by ruiann** [comment]
@heloowird
So it seems that the decode call will cost more and more time with the times of call increases?

@yzhwang

**Comment by heloowird** [comment]
yes, you can simply test time cost of `x_batch = [get_image(image.eval()) for i in xrange(batch_size)]`

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assigneee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yzhwang** [comment]
@ruiann Could you try converting image files to TFRecord first (example here: https://github.com/tensorflow/models/blob/master/research/inception/inception/data/build_image_data.py), then using our new Dataset abstraction? I believe that will solve the problem too.
https://www.tensorflow.org/programmers_guide/datasets

I haven't spent too much time on figuring out why tf.image.decode_image() takes longer and uses more memory as training steps go up. First could you please verify that the problem can be solved by either: 1) moving the batch building to graph building, or 2) trying to use Dataset?

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by ruiann** [comment]
@yzhwang 
thx for your reply. I'll make some try, but I think it's better to solve the tf.image.decode_image issue.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yzhwang** [comment]
@drpngx Do you know who would be a better one to handle this issue on tf.image.decode_image()?

**Comment by drpngx** [comment]
@ruiann could you run the heap profiler to see the call graph of the allocations?

**Comment by ruiann** [comment]
@drpngx 
I'm sorry but I'm not familiar with memory debug😂

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by gvashkevich** [comment]
Hi guys! I have the same problem. I have noticed that the training time is not increasing when I disable all summary operations.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yzhwang: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by wuchichung** [comment]
same problem here when using tensorflow object detection api's faster_rcnn_inception

**Comment by reedwm** [comment]
Reading the comments above, the problem seemed to be with this file: https://github.com/ruiann/SRGAN/blob/master/SRGAN/train.py. I don't know how to run that file, but it adds ops to the graph every iteration, by [calling get_image](https://github.com/ruiann/SRGAN/blob/cfeae6823b41bfcafe6610c423e921b4837ab1de/SRGAN/train.py#L59). Adding ops every iteration will increase memory usage, since ops take memory. I haven't tested, but perhaps it can also make each iteration take longer, since there are more ops in the graph which perhaps are iterated over.

@ruiann, I'm closing this issue for now. If you still have the problem after removing all op creations before the loop, feel free to reopen and/or comment. If anyone else has this problem, please file a new issue with a self-contained example.

**Comment by nilskk** [comment]
@wuchichung do you found a solution to your problem? I'm facing the same issue.
